pkg/sql/virtual_schema.go

```diff
@@ -385,108 +378,78 @@ func (e virtualDefEntry) getPlanInfo(
 			}
 
 			// Figure out the ordinal position of the column that we're filtering on.
-			// TODO(jordan): there is probably a better way to do this.
-			constraintColumnOrd := table.ColumnIdxMap()[index.ColumnIDs[0]]
-			addRowIfPassesFilter := func(datums ...tree.Datum) error {
-				validateRow(datums...)
-				matchedFilter := false
-				// If we have a constraint, apply it as a filter. If the filter passes,
-				// or we had no filter at all, add the datums to the row container.
-				datumToFilter := datums[constraintColumnOrd]
-				for i := 0; i < idxConstraint.Spans.Count(); i++ {
-					span := idxConstraint.Spans.Get(i)
-					if span.HasSingleKey(p.EvalContext()) {
-						matchedFilter = datumToFilter.Compare(p.EvalContext(), span.StartKey().Value(0)) == 0
-						if matchedFilter {
-							break
-						}
-					} else {
-						// Check both boundaries - the datum must be greater than the
-						// startKey, and less than the endKey.
-						var matchedStart, matchedEnd bool
-						if span.StartKey().IsEmpty() {
-							matchedStart = true
-						} else {
-							startCmp := datumToFilter.Compare(p.EvalContext(), span.StartKey().Value(0))
-							if span.StartBoundary() == constraint.IncludeBoundary {
-								matchedStart = startCmp >= 0
-							} else {
-								matchedStart = startCmp > 0
-							}
-						}
-						if span.EndKey().IsEmpty() {
-							matchedEnd = true
-						} else {
-							endCmp := datumToFilter.Compare(p.EvalContext(), span.EndKey().Value(0))
-							if span.EndBoundary() == constraint.IncludeBoundary {
-								matchedEnd = endCmp <= 0
-							} else {
-								matchedEnd = endCmp < 0
-							}
-						}
-						matchedFilter = matchedStart && matchedEnd
-						if matchedFilter {
-							break
-						}
-					}
-				}
+			columnIdxMap := table.ColumnIdxMap()
+			indexKeyDatums := make([]tree.Datum, len(index.ColumnIDs))
 
-				var err error
-				if matchedFilter {
-					_, err = v.rows.AddRow(ctx, datums)
+			generator, cleanup := setupGenerator(ctx, func(pusher rowPusher) error {
+				var span constraint.Span
+				addRowIfPassesFilter := func(datums ...tree.Datum) error {
+					validateRow(datums...)
+					for i, id := range index.ColumnIDs {
+						indexKeyDatums[i] = datums[columnIdxMap[id]]
+					}
+					// Construct a single key span out of the current row, so that
+					// we can test it for containment within the constraint span of the
+					// filter that we're applying. The results of this containment check
+					// will tell us whether or not to let the current row pass the filter.
+					key := constraint.MakeCompositeKey(indexKeyDatums...)
+					span.Init(key, constraint.IncludeBoundary, key, constraint.IncludeBoundary)
+					var err error
+					if idxConstraint.ContainsSpan(p.EvalContext(), &span) {
+						validateRow(datums...)
```

Comment:
It's kind of weird that validate row calls into log.Fatal rather than just returning an error.